subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/DefaultDependencyConstraintSet.java

```diff
@@ -46,7 +43,7 @@ public String toString() {
     @Override
     public boolean add(final DependencyConstraint dependencyConstraint) {
         assertConfigurationIsDeclarableAgainst();
-        warnIfConfigurationIsDeprecated();
+        clientConfiguration.maybeEmitDeclarationDeprecation();
```

Comment:
💭 Kind of odd seeing these calls next to each other.  Should `assertConfigurationIsDeclarableAgainst()` be moved to `ConfigurationInternal` as well?  There's a similar method on `DefaultConfiguration`, this could consolidate them into a single `default` method.

Reply:
Why is it odd to have these next to each other? We want to throw an exception if its disabled for declaration and warn if its deprecated for declaration. It seems valid to me

Reply:
I thought it read a little odd that one is accessing a configuration to read its state and the other is asking the configuration to do something based on its state rather than both just calling methods on the configuration.

Reply:
I see what you mean. I'll see if I can clean this up a little

Reply:
Really what we need is a unified system to merge the existing DeprecationLogger deprecations used here and the exception we're throwing here, with the changes you just merged for deprecating methods per-role. Choosing when to throw exceptions and when to warn should all be consolidated in one place per operation we're guarding. 

I think its best to push this to a follow-up PR.